src/segment-loader.js

```diff
@@ -0,0 +1,400 @@
+
+      this.abort_();
+
+      // don't wait for buffer check timeouts to begin fetching the
+      // next segment
+      if (!this.paused()) {
+        this.state = 'READY';
+        this.fillBuffer_();
+      }
+    },
+    error: function(error) {
+      if (error !== undefined) {
+        this.error_ = error;
+      }
+
+      this.pendingSegment_ = null;
+      return this.error_;
+    },
+    load: function() {
+      this.monitorBuffer_();
+
+      // if we don't have a playlist yet, keep waiting for one to be
+      // specified
+      if (!this.playlist_) {
+        return;
+      }
+
+      // if we're in the middle of processing a segment already, don't
+      // kick off an additional segment request
+      if (this.state !== 'READY' && this.state !== 'INIT') {
+        return;
+      }
+
+      this.state = 'READY';
+      this.fillBuffer_();
+    },
+    playlist: function(media) {
+      this.playlist_ = media;
+
+      // if we were unpaused but waiting for a playlist, start
+      // buffering now
+      if (media && this.state === 'INIT' && !this.paused()) {
+        this.state = 'READY';
+        return this.fillBuffer_();
+      }
+    },
+    /**
+     * Prevent the loader from fetching additional segments. If there
+     * is a segment request outstanding, it will finish processing
+     * before the loader halts. A segment loader can be unpaused by
+     * calling load().
+     */
+    pause: function() {
+      if (this.checkBufferTimeout_) {
+        window.clearTimeout(this.checkBufferTimeout_);
+
+        this.checkBufferTimeout_ = null;
+      }
+    },
+    /**
+     * Returns whether the segment loader is fetching additional
+     * segments when given the opportunity. This property can be
+     * modified through calls to pause() and load().
+     */
+    paused: function() {
+      return this.checkBufferTimeout_ === null;
+    },
+    timestampOffset: function(offset) {
+      this.timestampOffset_ = offset;
+    },
+
+    monitorBuffer_: function() {
+      if (this.state === 'READY') {
+        this.fillBuffer_();
+      }
+      this.checkBufferTimeout_ = window.setTimeout(this.monitorBuffer_.bind(this),
+                                                   CHECK_BUFFER_DELAY);
+    },
+    /**
+     * Determines what segment request should be made, given current
+     * playback state.
+     * @param buffered {TimeRanges} the state of the buffer
+     * @param playlist {object} the playlist object to fetch segments from
+     * @param currentTime {number} the playback position in seconds
+     * @param timestampOffset (optional} {number} the duration of
+     * content, in seconds, that has been removed from the front of
+     * this playlist. If unspecified, it is assumed to be zero.
+     * @return {object} a segment info object that describes the
+     * request that should be made or null if no request is necessary
+     */
+    checkBuffer_: function(buffered, playlist, currentTime, timestampOffset) {
```

Comment:
I intentionally made this function explicit in the inputs it accepts, instead of relying on state stored on `this` (maybe I should have put it on the module object instead of making it instance level), so that we could easily unit test it. Hopefully this will make tracking down stream-sync issues easier to debug.